sw/device/silicon_creator/lib/drivers/usb.c

```diff
@@ -0,0 +1,393 @@
+  uint8_t id = (uint8_t)bitfield_find_first_set32((int32_t)buffer_pool) - 1;
+  buffer_pool &= ~(1 << id);
+  return id;
+}
+
+static inline bool buffer_pool_empty(void) { return buffer_pool == 0; }
+
+usb_ep_info_t endpoints[USBDEV_PARAM_N_ENDPOINTS];
+
+// Copy memory into a USB buffer.
+// The USB register space only permits word writes, so we need to handle
+// buffer lengths that are not perfect multiples of the word size.
+static void copy_to_buffer(uint8_t id, const char *src, size_t len) {
+  uintptr_t buffer = kBase + USBDEV_BUFFER_REG_OFFSET + (uint32_t)id * 64;
+  volatile uint32_t *dst = (volatile uint32_t *)buffer;
+  while (len >= sizeof(uint32_t)) {
+    *dst++ = *(uint32_t *)src;
+    src += sizeof(uint32_t);
+    len -= sizeof(uint32_t);
+  }
+  uint32_t last = 0;
+  uint32_t shift = 0;
+  while (len > 0) {
+    last |= (uint32_t)*src << shift;
+    shift += 8;
+    src += 1;
+    len -= 1;
+  }
+  *dst = last;
+}
+
+// Copy from an USB buffer to memory.
+// The USB register space only permits word writes, so we need to handle
+// buffer lengths that are not perfect multiples of the word size.
+static void copy_from_buffer(uint8_t id, char *dst, size_t len) {
+  uintptr_t buffer = kBase + USBDEV_BUFFER_REG_OFFSET + (uint32_t)id * 64;
+  volatile uint32_t *src = (volatile uint32_t *)buffer;
+
+  uint32_t *dst32 = (uint32_t *)dst;
+  while (len >= sizeof(uint32_t)) {
+    *dst32++ = *src++;
+    len -= sizeof(uint32_t);
+  }
+  if (len > 0) {
+    uint32_t last = *src;
+    dst = (char *)dst32;
+    while (len > 0) {
+      *dst++ = (char)last;
+      last >>= 8;
+      len -= 1;
+    }
+  }
+}
+
+// Configure the PHY according the define diretives above.
+static void usb_phy_init(void) {
+  uint32_t phy_config = 0;
+  phy_config = bitfield_bit32_write(
+      phy_config, USBDEV_PHY_CONFIG_USE_DIFF_RCVR_BIT, CFG_EN_DIFF_RCVR);
+  phy_config = bitfield_bit32_write(
+      phy_config, USBDEV_PHY_CONFIG_TX_USE_D_SE0_BIT, CFG_TX_USE_D_SE0);
+  phy_config = bitfield_bit32_write(
+      phy_config, USBDEV_PHY_CONFIG_EOP_SINGLE_BIT_BIT, false);
+  phy_config = bitfield_bit32_write(phy_config, USBDEV_PHY_CONFIG_PINFLIP_BIT,
+                                    CFG_PIN_FLIP);
+  phy_config = bitfield_bit32_write(
+      phy_config, USBDEV_PHY_CONFIG_USB_REF_DISABLE_BIT, true);
+  abs_mmio_write32(kBase + USBDEV_PHY_CONFIG_REG_OFFSET, phy_config);
+}
+
+// Performa a read/modify/write on a given register.
+static void usbreg_bit(uint32_t offset, uint32_t bit, bool value) {
+  uint32_t reg = abs_mmio_read32(kBase + offset);
+  reg = bitfield_bit32_write(reg, bit, value);
+  abs_mmio_write32(kBase + offset, reg);
+}
+
+// Set or clear STALL on an endpoint.
+void usb_ep_stall(size_t ep, bool enable) {
+  if (endpoints[ep].type & kUsbEndpointTypeIn) {
+    usbreg_bit(USBDEV_IN_STALL_REG_OFFSET, ep, enable);
+  }
+  if (endpoints[ep].type & kUsbEndpointTypeOut) {
+    usbreg_bit(USBDEV_OUT_STALL_REG_OFFSET, ep, enable);
+  }
```

Comment:
Generally-speaking, these STALL settings can happen independently. This is changing two different endpoints potentially at once.

I suppose we could say there's a firmware limitation that artificially limits how many of the 24 endpoints can be used, though.

Reply:
Fixed.